erpnext/stock/stock_ledger.py

```diff
@@ -65,22 +69,67 @@ def make_sl_entries(sl_entries, allow_negative_stock=False, via_landed_cost_vouc
+			sle = frappe._dict(sle)
+			voucher_doc = frappe.get_doc(sle.voucher_type, sle.voucher_no)
+
+			if sle.get("actual_qty") and item_det.has_serial_no and sle.get("voucher_type")!="Stock Reconciliation":
+				# for auto creating serials
+				if not via_landed_cost_voucher:
+					validate_sr_no(sle, item_det)
+				if not sle.serial_no and item_det.has_serial_no and (not voucher_doc.is_return or
+					(voucher_doc.doctype == "Sales Invoice")):
+					sle = update_serial_nos(sle, item_det, return_sle=True)
+
+					for item in voucher_doc.get('items'):
+						if item.get('item_code') == sle.item_code:
+							if sle.get("voucher_type")=="Stock Entry":
+								if (item.s_warehouse or item.t_warehouse) == sle.warehouse:
+									frappe.db.set_value(item.doctype, item.name, 'serial_no', sle.serial_no)
+									break
+							if item.warehouse == sle.warehouse:
+								frappe.db.set_value(item.doctype, item.name, 'serial_no', sle.serial_no)
+							elif item.rejected_warehouse == sle.warehouse:
+								frappe.db.set_value(item.doctype, item.name, 'rejected_serial_no', sle.serial_no)
+							break
+					serials = sle.serial_no.strip().split('\n')
+					for serial in range(abs(cint(sle['actual_qty']))):
+						new_sle = sle.copy()
+						new_sle['actual_qty'] = 1 if sle['actual_qty'] > 0 else -1
+						new_sle['serial_no'] = serials.pop()
+						new_sle['skip_update_serial_no'] = 1
+						new_sle['skip_serial_no_validation'] = 1
+						sle_docs.append(make_entry(new_sle, allow_negative_stock, via_landed_cost_voucher))
+						sr = frappe.get_doc('Serial No', new_sle['serial_no'])
+						sr.update_serial_no_reference(sr.serial_no, update=True)
+
+				# for manually entered serials
+				elif sle.serial_no:
+					serials = sle.serial_no.strip().split('\n')
+					for serial in range(abs(cint(sle['actual_qty']))):
+						new_sle = sle.copy()
+						new_sle['actual_qty'] = 1 if sle['actual_qty'] > 0 else -1
+						new_sle['serial_no'] = serials.pop()
+						new_sle['skip_serial_no_validation'] = 1
+						sle_docs.append(make_entry(new_sle, allow_negative_stock, via_landed_cost_voucher))
+
+			elif sle.get("actual_qty") or sle.get("voucher_type")=="Stock Reconciliation":
+				sle_docs.append(make_entry(sle, allow_negative_stock, via_landed_cost_voucher))
+
+			for sle_doc in sle_docs:
+				args = sle_doc.as_dict()
+
+				if sle.get("voucher_type") == "Stock Reconciliation":
+					# preserve previous_qty_after_transaction for qty reposting
+					args.previous_qty_after_transaction = sle.get("previous_qty_after_transaction")
+
+				is_stock_item = frappe.get_cached_value('Item', args.get("item_code"), 'is_stock_item')
+				if is_stock_item:
+					bin_name = get_or_make_bin(args.get("item_code"), args.get("warehouse"))
+					repost_current_voucher(args, allow_negative_stock, via_landed_cost_voucher)
+					update_bin_qty(bin_name, args)
+				else:
+					frappe.msgprint(_("Item {0} ignored since it is not a stock item").format(args.get("item_code")))
```

Comment:
This function is now 100 lines long. All serial splitting logic can be easily moved to a separate function.